File: frontend/src/types/bulletins.ts
export interface StatsPedagogiques {
  classe_id: string;
  classe_nom: string;
  periode_libelle: string;
  effectif: number;
  moyenne_classe: number | null;
  taux_reussite: number | null;
  meilleur_eleve: string | null;
  moins_bonne_moyenne: number | null;
}

export interface PalmaresItem {
  rang: number;
  eleve_id: string;
  nom: string;
  prenoms: string;
  matricule: string;
  moyenne_generale: number;
  appreciation: string | null;
}

export interface Palmares {
  classe_id: string;
  classe_nom: string;
  periode_libelle: string;
  items: PalmaresItem[];
}

export interface DecisionPassage {
  id: string;
  eleve_id: string;
  annee_scolaire_id: string;
  decision: string;
  observation: string | null;
}

export interface Competence {
  id: string;
  code: string;
  domaine: string;
  libelle: string;
  niveau_id: string;
  ordre: number;
}

export interface CompetenceGrilleRow {
  eleve_id: string;
  matricule: string;
  nom: string;
  prenoms: string;
  evaluations: Record<string, string | null>;
}

export interface CompetenceGrille {
  classe_id: string;
  classe_nom: string;
  periode_libelle: string;
  competences: Competence[];
  eleves: CompetenceGrilleRow[];
}

export type StatutCompetence = "acquis" | "en_cours" | "non_acquis" | "";

export type DecisionPassageType = "admis" | "redouble" | "exclu";

File: backend/app/schemas/bulletins.py
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.notes import EleveMoyennesItem


class BulletinEleveData(BaseModel):
    eleve_id: UUID
    matricule: str
    nom: str
    prenoms: str
    classe_nom: str
    periode_libelle: str
    annee_libelle: str
    moyennes: EleveMoyennesItem
    decision: str | None = None
    observation: str | None = None


class StatsPedagogiquesResponse(BaseModel):
    classe_id: UUID
    classe_nom: str
    periode_libelle: str
    effectif: int
    moyenne_classe: Decimal | None
    taux_reussite: Decimal | None
    meilleur_eleve: str | None
    moins_bonne_moyenne: Decimal | None


class PalmaresItem(BaseModel):
    rang: int
    eleve_id: UUID
    nom: str
    prenoms: str
    matricule: str
    moyenne_generale: Decimal
    appreciation: str | None


class PalmaresResponse(BaseModel):
    classe_id: UUID
    classe_nom: str
    periode_libelle: str
    items: list[PalmaresItem]


class DecisionPassageCreate(BaseModel):
    eleve_id: UUID
    decision: str = Field(..., pattern=r"^(admis|redouble|exclu)$")
    observation: str | None = None
    annee_scolaire_id: UUID | None = None


class DecisionPassageResponse(BaseModel):
    id: UUID
    eleve_id: UUID
    annee_scolaire_id: UUID
    decision: str
    observation: str | None

    model_config = {"from_attributes": True}


class CompetenceResponse(BaseModel):
    id: UUID
    code: str
    domaine: str
    libelle: str
    niveau_id: UUID
    ordre: int

    model_config = {"from_attributes": True}


class CompetenceEvalItem(BaseModel):
    competence_id: UUID
    statut: str = Field(..., pattern=r"^(acquis|en_cours|non_acquis)$")
    appreciation: str | None = None


class CompetenceBulkUpdate(BaseModel):
    eleve_id: UUID
    evaluations: list[CompetenceEvalItem]


class CompetenceGrilleRow(BaseModel):
    eleve_id: UUID
    matricule: str
    nom: str
    prenoms: str
    evaluations: dict[str, str | None]


class CompetenceGrilleResponse(BaseModel):
    classe_id: UUID
    classe_nom: str
    periode_libelle: str
    competences: list[CompetenceResponse]
    eleves: list[CompetenceGrilleRow]
